Update behavior for postCommand.strategy=post-index-change

derrickstolee opened the pull request:
This feature was introduced in #736 but had a few pain points that were discovered when testing the full replacement of the existing post-index-change and post-command hooks in our internal environment:

1. When using `core.hooksPath` to point to a directory within the worktree, the sentinel files were being written into the worktree. Write to `$GITDIR/info/` instead.
2. The existing internal post-index-change hook only signaled that work needed to be done if actually the _worktree_ changed. Make the sentinel file logic more restrictive to only write in these cases.

Tests are added to confirm both of these behavior changes.

derrickstolee committed:
fixup! hooks: add custom post-command hook config

This change to use the 'info/' directory instead of the 'hooks/'
directory is important because the hooks dir could be redirected to the
source tree.

Modify the test in such a way that this code change would fail due to
being unable to write to the internal hooks directory.

Signed-off-by: Derrick Stolee <stolee@gmail.com>

derrickstolee committed:
hook: extract the internal hook replacement logic

Before making the post-index-change logic more advanced for the
postCommand.strategy=post-index-change feature, extract the code into a
new method that will be easier to update in the future.

Signed-off-by: Derrick Stolee <stolee@gmail.com>

derrickstolee committed:
hook: restrict post-command hook even more

The postCommand.strategy=post-index-change config was designed to
replace some internal hook usage. This internal hook was more
restrictive than the initial feature was designed. Specifically, it
ignored index updates that did not update the worktree. With the
existing implementation, this causes more work done in the post-command
hook than intended.

Update the code to write a sentinel file only when the worktree is
changed.

Signed-off-by: Derrick Stolee <stolee@gmail.com>

dscho commented on Documentation/config/postcommand.adoc:
```diff
@@ -9,5 +9,5 @@ postCommand.strategy::
 `post-index-change`;;
 	run the `post-command` hook only if the current process wrote to
-	the index.
+	the index and updated the worktree.
```
I don't mind this change, even if the name `post-index-change` does not quite convey that index changes that do not result in worktree changes will be ignored. But I cannot think of a better name, can you?

derrickstolee replied:
We could `s/post-index-change/worktree-change/` as an option.

derrickstolee replied:
The biggest reason to leave the name is that this strategy _replaces_ the `post-index-change` hook.

dscho replied:
I do like `worktree-change`!

While the history makes `post-index-change` a natural choice, the `postCommand.strategy` setting does not _disable_ or _override_ any `post-index-change` hooks, they will still be executed if they exist.

Unless you deem it too much work for the benefit, would you mind renaming it?

derrickstolee replied:
I can rename it, but I do want to point out that it _does_ override the `post-index-change` hook if it exists (we exit early from `run_hooks_opt()` if it is trying to run the `post-index-change` hook). I can change that behavior if you need it.

dscho replied:
Oh... Okay, let's leave it then. Do we have any elegant way out in case a user _does_ require the `post-command` to be run on index changes _including_ ones outside the sparse cone?

derrickstolee replied:
My latest commit hopes to clarify this:

* The point of this feature is to skip running the post-command hook if the worktree didn't change. This helps avoid overhead that is wasting time, especially for should-be-fast Git commands.
* In our internal use case, this means that we will no longer need our post-index-change hook. (Thus, there is no reason to skip it.)
* The new value `postCommand.strategy=worktree-change` helps to signal that we are not interrupting any post-index-change hooks but instead running the post-command hook in only certain cases.

dscho reviewed (dismissed):
Makes sense, and since the feature is young enough, I have no objection to that minor behavior change.

derrickstolee committed:
hooks: change postCommand.strategy config value

The previously-named 'post-index-change' value for
'postCommand.strategy' did the following two things:

 1. Avoid running the post-command hook unless the post-index-change
    hook _would_ run with a signal that the worktree changed.

 2. Avoid running any installed post-index-change hooks.

The additional restriction of the worktree change in (1) makes this name
less appropriate. Also, item (2) seems to be a side-effect that we
should avoid. We would like to allow both behaviors to exist.

Update the value and behavior, including the tests.

Signed-off-by: Derrick Stolee <stolee@gmail.com>

derrickstolee committed:
hooks: simplify handle_hook_replacement()

Now that we don't skip the post-index-change hook, we don't need to pass
the return value as an out-parameter in this method.

Signed-off-by: Derrick Stolee <stolee@gmail.com>
